### PR TITLE
Edit profile

### PR DESCRIPTION
**Pull request by sector46**
This is a pull request for #13 and #26. Allows user to edit their profile.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -16,12 +16,13 @@ class User < ActiveRecord::Base
     uniqueness: { case_sensitive: false }
 
   has_secure_password
-  validates :password, length: { minimum: 6 }
+  validates :password, length: { minimum: 6 }, allow_nil: true
```
Comment (lewisric) [inline app/models/user.rb]: Maybe im dumb, but are we dictating password length and allowing for null passwords here?

Comment (lewisric) [inline app/models/user.rb]: disregard I see what it's doing

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -18,6 +19,51 @@ def create
...
+    # the edit view will return 'false'
+    if params[:changing_password] == 'false'
+      params[:user].delete :password
+      params[:user].delete :password_confirmation
+      if @user.update_attributes(user_params)
+        redirect_to @user
+      else
+        # rendering edit will change URL, possibly
+        # breaking button to 'Change password' in 'edit' view
+        # FIXME: correct URL in later version
```
Comment (lewisric) [inline app/controllers/users_controller.rb]: We should figure out if this is an issue sooner rather than later

**Comment by lewisric** [comment]
@CATcommander @beazley2 @ardeshirj 

Please do a CR

**Comment by ardeshirj** [comment]
Can you just briefly tell us how this works?

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (beazley2) [inline app/controllers/users_controller.rb]: This does break the change password button if you try to update your profile with bogus info.

**Comment by beazley2** [comment]
Looks, except there's no protection from users updating someone else's profile. https://www.railstutorial.org/book/updating_and_deleting_users shows how to incorporate that.

**Comment by sector46** [comment]
So the way that this works is that there are two views for updating a user. 1) the 'edit' view updates all user info except the password. 2) the 'change_password view updates the password by verifying the current password and then accepts a new password (after validation). The way the 'update' action differentiates between these views is from the :changing_password param that comes from the views themselves. I'm going to update the code so that users are protected from having someone else updating their profile and that each attribute is validated/updated separately (though this may cause us to lose form error responses on these forms, I'll have to see if I can update that).

**Comment by ardeshirj** [comment]
Thanks for the explaining it. 

Lets do a demo in next meeting, so we all get in same page with you before we actually merge it to master.

**Comment by lewisric** [comment]
agreed

**Comment by sector46** [comment]
To reproduce URL rendering error: 
1) Login as some user
2) Enter 
http://0.0.0.0:3000/users/<Insert user ID here>/edit 
as URL. <Insert user ID> can be found by logging in and clicking on Account->Profile.
3) Erase any of the populated information fields (Ex: change "Name: Test" to "Name: "
4) Submit the information by clicking "Save changes".
5) You should now see a validation error.
6) Click on "Change password" to attempt to change password.
7) Error "ActiveRecord::RecordNotFound in UsersController#show" should be reproduced.

This is caused by the "render(view)" method, which changes the URL to match the "user" controller and "show" action when it runs "render".

**Comment by ardeshirj** [comment]
The new commits should resolve the issue #40. Ready for review.

**Comment by beazley2** [comment]
Review complete. No errors found.

**Comment by lewisric** [comment]
CR complete

**Pull request merged**